Core/FitParameter.hpp

```diff
@@ -1,230 +1,67 @@
-
-  /// Setter parameter error and assume that this parameter has symmetric
-  /// errors.
-  virtual void setError(double err);
-
-  bool operator==(const FitParameter otherPar) const;
-
-  /// Load parameters from a ptree. This approach is more or
-  /// less equivalent to the serialization of a parameter but provides a better
-  /// readable format.
-  void load(const boost::property_tree::ptree &pt);
-
-  /// Save parameter to a ptree. This approach is more or
-  /// less equivalent to the serialization of a parameter but provides a better
-  /// readable format.
-  boost::property_tree::ptree save() const;
-
-  /// String with detailed information about the parameter. Used in
-  /// operator<<().
-  virtual std::string to_str() const;
-
-  /// String with detailed information about the parameter. Used in
-  /// operator<<().
-  virtual std::string val_to_str() const;
-
-protected:
-  virtual std::string className() const { return "Double"; }
-
-  /// Are valid bounds defined for this parameter?
-  bool HasBounds;
-
-  /// Do you want to keep parameter fixed?
-  bool IsFixed;
-
-  /// Parameter value
-  double Value;
-
-  /// Parameter bounds
-  std::pair<double, double> Bounds;
-
-  /// No error / symmetric error / asymmetric error
-  ErrorType ErrType;
-
-  /// Lower parameter error
-  std::pair<double, double> Error;
-
-  virtual void SetErrorType(ErrorType t) { ErrType = t; }
-
-  /// Check if \p min and \p max are valid bounds
-  bool check_bounds(const std::pair<double, double> bounds) const;
-
-private:
-  friend class boost::serialization::access;
-  template <class archive>
-  void serialize(archive &ar, const unsigned int version) {
-    using namespace boost::serialization;
-    //    ar &boost::serialization::make_nvp(
-    //        "Parameter", boost::serialization::base_object<Parameter>(*this));
-    ar &make_nvp("Name", Name);
-    ar &make_nvp("Bounds", Bounds);
-    ar &make_nvp("Fix", IsFixed);
-    ar &make_nvp("Value", Value);
-    ar &make_nvp("Bounds", Bounds);
-    try {
-      ar &make_nvp("ErrorType", ErrType);
-      ar &make_nvp("Error", Error);
-    } catch (...) {
-      Error = std::pair<double, double>(0, 0);
-      ErrType = ErrorType::SYM;
+template <typename T> struct FitParameter {
+  FitParameter() = default;
+  FitParameter(std::string name, T val, bool isfixed = true)
+      : HasBounds(false), IsFixed(isfixed), Value(val), Name(name),
+        Bounds(0.0, 0.0) {}
+  FitParameter(std::string name, T val, T min, T max, bool isfixed = true)
+      : HasBounds(true), IsFixed(isfixed), Value(val), Name(name),
+        Bounds(min, max) {}
+  bool HasBounds = false;
+  bool IsFixed = false;
+  T Value;
+  std::pair<T, T> Error;
+  std::string Name = "";
+  std::pair<T, T> Bounds;
+
+  friend std::ostream &operator<<(std::ostream &os,
+                                  const FitParameter<double> &x) {
+    os << x.Name << ": " << x.Value;
+    if (x.Error.first != 0.0 || x.Error.second != 0.0) {
```

Comment:
Is this a good way to check if some double variable is equal to a specific value? Or it is a comparison using !=, so we can use it safely?

Reply:
Hmm. Yeah, since I simply want to know if errors exist, the best would be to use `std::optional` ( which comes with c++17). I can also introduce a bool just like for the bounds. But sooner or later we should upgrade the bound and error to std::optionals.